src/Engines/Native/nativeDataStream.ts

```diff
@@ -0,0 +1,115 @@
+
+        this._position = 0;
+
+        this._nativeDataStream = new _native.NativeDataStream(() => {
+            this._flush();
+        });
+    }
+
+    public writeUint8(value: number): void {
+        this._autoFlushIfNecessary(1);
+        this._data.setUint8(this._position, value);
+        this._position += 1;
+    }
+
+    public writeUint32(value: number): void {
+        this._autoFlushIfNecessary(4);
+        this._data.setUint32(this._position, value, true);
+        this._position += 4;
+    }
+
+    public writeInt32(value: number): void {
+        this._autoFlushIfNecessary(4);
+        this._data.setInt32(this._position, value, true);
+        this._position += 4;
+    }
+
+    public writeFloat32(value: number): void {
+        this._autoFlushIfNecessary(4);
+        this._data.setFloat32(this._position, value, true);
+        this._position += 4;
+    }
+
+    public writeUint32Array(values: Uint32Array): void {
+        this._byteAlign32();
+        this._autoFlushIfNecessary(4 + values.byteLength);
+        let index = NativeDataStream._positionToIndex(this._position);
+        this._uint32s[index] = values.length;
+        this._uint32s.set(values, index + 1);
+        this._position = NativeDataStream._indexToPosition(index + values.length + 1);
+    }
+
+    public writeInt32Array(values: Int32Array): void {
+        this._byteAlign32();
+        this._autoFlushIfNecessary(4 + values.byteLength);
+        let index = NativeDataStream._positionToIndex(this._position);
+        this._uint32s[index] = values.length;
+        this._int32s.set(values, index + 1);
+        this._position = NativeDataStream._indexToPosition(index + values.length + 1);
+    }
+
+    public writeFloat32Array(values: Float32Array): void {
+        this._byteAlign32();
+        this._autoFlushIfNecessary(4 + values.byteLength);
+        let index = NativeDataStream._positionToIndex(this._position);
+        this._uint32s[index] = values.length;
+        this._float32s.set(values, index + 1);
+        this._position = NativeDataStream._indexToPosition(index + values.length + 1);
+    }
+
+    public writeNativeData(handle: NativeData) {
+        this.writeUint32Array(handle);
+    }
+
+    public writeBoolean(value: boolean) {
+        this.writeUint32(value ? 1 : 0);
+    }
+
+    private _byteAlign32(): void {
+        this._position = NativeDataStream._indexToPosition(NativeDataStream._positionToIndex(this._position));
+    }
+
+    private _autoFlushIfNecessary(requiredBytes: number): void {
+        if (this._position + requiredBytes > this._data.byteLength) {
+            this._flush();
+        }
+    }
+
+    private _flush(): void {
+        this._nativeDataStream.writeBytes(this._data.buffer, this._position);
+        this._position = 0;
+    }
+
+    private static _positionToIndex(position: number): number {
+        return position % 4 === 0 ? position / 4 : Math.floor(position / 4) + 1;

```

Comment:
there must be a simpler way to do this...

Reply:
I think what I originally considered was something like `Math.floor((position - 1) / 4) + 1`, which is simpler in some ways, but I switched to this because I thought the other was harder to read and understand why it was doing what it was doing. Don't have a real preference between them, so happy to switch it back if you think the smaller way is better.